### PR TITLE
perf: api connector

### PR DESCRIPTION
**Pull request by wenytang-ms**
Add api-connector to fx-core as a plugin
1. add command pallete as ' teams: Connect to an API'
2. add user question
3. generate sample code
4. update .env.teamsfx.local file

E2E TEST: https://github.com/OfficeDev/TeamsFx/actions/runs/2037961347

### CODE REVIEW

**Comment by wenytang-ms** [comment]
E2E TEST: n/a

File: packages/fx-core/src/plugins/resource/apiconnector/result.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+"use strict";
+import { FxError, SystemError, UserError, Result, ok } from "@microsoft/teamsfx-api";
```
Comment (xzf0587) [inline packages/fx-core/src/plugins/resource/apiconnector/result.ts:4]: Just a reminder. The error construct has changed in hotfix branch.
We should update this after the dev merged hotfix branch.

Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/result.ts:4]: thanks

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      options.push(functionOption);
+    }
+    if (options.length === 0) {
+      throw ResultFactory.UserError("no bot plugin or func plugin", "no bot plugin or func plugin");
+    }
+    const whichService = new QTreeNode({
+      name: Constants.questionKey.serviceSelect,
+      type: "singleSelect",
+      staticOptions: options,
+      title: getLocalizedString("plugins.apiConnector.whichService.title"),
```
Comment (blackchoey) [inline packages/fx-core/src/plugins/resource/apiconnector/index.ts:74]: refine the wording of `service`

File: packages/vscode-extension/src/extension.ts [hunk 3]
```diff
@@ -156,6 +156,11 @@ export async function activate(context: vscode.ExtensionContext) {
   );
   context.subscriptions.push(validateManifestCmd);
 
+  const addApiConnector = vscode.commands.registerCommand(
+    "fx-extension.addApiConnector",
+    (...args) => Correlator.run(handlers.addApiConnectorHandler, args)
+  );
```
Comment (blackchoey) [inline packages/vscode-extension/src/extension.ts:162]: add feature flag in another pr

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts (hunk 2 above)
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts]: sure, replease with component

File: packages/vscode-extension/src/extension.ts [hunk 4]
```diff
@@ -156,6 +156,11 @@ export async function activate(context: vscode.ExtensionContext) {
   );
   context.subscriptions.push(validateManifestCmd);
 
+  const addApiConnector = vscode.commands.registerCommand(
+    "fx-extension.addApiConnector",
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/vscode-extension/src/extension.ts:160]: sure

File: packages/fx-core/src/plugins/resource/apiconnector/plugin.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    const fileSuffix: string = languageType === LanguageType.JS ? FileType.JS : FileType.TS;
+    const sampleCodeDirectory = path.join(
+      getTemplatesFolder(),
+      "plugins",
+      "resource",
+      "apiconnector",
+      "sample",
+      fileSuffix
+    );
+    const fileName: string = Constants.pluginNameShort + "." + fileSuffix;
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/plugin.ts]: sure

File: packages/fx-core/src/plugins/resource/apiconnector/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    if (!inputs.projectPath) {
+      throw ResultFactory.SystemError(
+        ErrorMessage.ApiConnectorPathError.name,
+        ErrorMessage.ApiConnectorPathError.message("")
+      );
+    }
+    const projectPath = inputs.projectPath;
+    const config: ApiConnectorConfiguration = this.getUserDataFromInputs(inputs);
+    const service: ProjectType =
+      config.ServicePath === ProjectType.BOT ? ProjectType.BOT : ProjectType.API;
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/plugin.ts]: updated

File: packages/fx-core/src/plugins/resource/apiconnector/plugin.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import { ApiConnectorResult, ResultFactory } from "./result";
+import { getTemplatesFolder } from "../../../folder";
+import { EnvHandler } from "./envHandler";
+import { ErrorMessage } from "./errors";
+export class ApiConnectorImpl {
+  public async scaffold(ctx: Context, inputs: Inputs): Promise<ApiConnectorResult> {
+    if (!inputs.projectPath) {
+      throw ResultFactory.SystemError(
+        ErrorMessage.ApiConnectorPathError.name,
+        ErrorMessage.ApiConnectorPathError.message("")
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/plugin.ts:32]: ok

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      type: "singleSelect",
+      staticOptions: [
+        BasicAuthOption,
+        CertAuthOption,
+        AADAuthOption,
+        APIKeyAuthOption,
+        OtherAuthOPtion,
+      ],
+      title: getLocalizedString("plugins.apiConnector.whichAuthType.title"),
+      default: BasicAuthOption.id,
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts]: updated.

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    });
+    const whichAuthType = new QTreeNode({
+      name: Constants.questionKey.apiType,
+      type: "singleSelect",
+      staticOptions: [
+        BasicAuthOption,
+        CertAuthOption,
+        AADAuthOption,
+        APIKeyAuthOption,
+        OtherAuthOPtion,
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts]: sure

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    if (options.length === 0) {
+      throw ResultFactory.UserError("no bot plugin or func plugin", "no bot plugin or func plugin");
+    }
+    const whichService = new QTreeNode({
+      name: Constants.questionKey.serviceSelect,
+      type: "singleSelect",
+      staticOptions: options,
+      title: getLocalizedString("plugins.apiConnector.whichService.title"),
+      default: options[0].id,
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts]: updated.

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    if (activeResourcePlugins.includes(ResourcePlugins.Function)) {
+      options.push(functionOption);
+    }
+    if (options.length === 0) {
+      throw ResultFactory.UserError("no bot plugin or func plugin", "no bot plugin or func plugin");
+    }
+    const whichService = new QTreeNode({
+      name: Constants.questionKey.serviceSelect,
+      type: "singleSelect",
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts]: updated.

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    ctx: Context,
+    inputs: Inputs,
+    func: Func,
+    envInfo: DeepReadonly<v2.EnvInfoV2>,
+    tokenProvider: TokenProvider
+  ): Promise<ApiConnectorResult> {
+    const activeResourcePlugins = (ctx.projectSetting.solutionSettings as AzureSolutionSettings)
+      ?.activeResourcePlugins;
+    if (!activeResourcePlugins) {
+      throw ResultFactory.UserError("no plugins", "no plugins");
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts:45]: sure

File: packages/fx-core/src/plugins/resource/apiconnector/index.ts [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    localSettings: Json,
+    envInfo: v2.EnvInfoV2,
+    tokenProvider: TokenProvider
+  ): Promise<ApiConnectorResult> {
+    if (func.method != "addApiConnector") {
+      return err(
+        new SystemError(
+          Constants.PLUGIN_NAME,
+          "FunctionRouterError",
+          `Failed to route function call:${JSON.stringify(func)}`
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/index.ts:64]: updated.

File: packages/vscode-extension/src/telemetry/extTelemetryEvents.ts [hunk 14]
```diff
@@ -40,6 +40,9 @@ export enum TelemetryEvent {
   UpdatePreviewManifestStart = "update-preview-manifest-start",
   UpdatePreviewManifest = "update-preview-manifest",
 
+  ConnectExistingApiStart = "connect-exist-api-start",
+  ConnectExistingApi = "aconnect-exist-api",
```
Comment (blackchoey) [inline packages/vscode-extension/src/telemetry/extTelemetryEvents.ts:44]: typo, by the way, can we name the event as connect-existing-api?

File: packages/fx-core/src/plugins/resource/apiconnector/sampleHandler.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  private readonly projectRoot: string;
+  private readonly laguageType: FileType;
+  private readonly component: string;
+  constructor(projectPath: string, languageType: string, component: string) {
+    this.projectRoot = projectPath;
+    this.laguageType = languageType === LanguageType.JS ? FileType.JS : FileType.TS;
+    this.component = component;
+  }
+
+  private getFileType(): FileType {
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/fx-core/src/plugins/resource/apiconnector/sampleHandler.ts]: done

File: packages/vscode-extension/src/telemetry/extTelemetryEvents.ts (hunk 14 above)
Comment (wenytang-ms) [reply packages/vscode-extension/src/telemetry/extTelemetryEvents.ts]: thanks!

**Review by blackchoey — DISMISSED** [review]

**Review by blackchoey — DISMISSED** [review]

**Review by jayzhang — APPROVED** [review]

File: packages/vscode-extension/src/resources/Strings.json [hunk 16]
```diff
@@ -39,6 +39,7 @@
       "validateManifestTitle": "Validate App Manifest File",
       "validateManifestTitleNew": "Validate manifest file",
       "validateManifestDescription": "Validate the Teams app manifest file",
+      "addApiConnectorTitle": "Connect to an API",
```
Comment (tecton) [inline packages/vscode-extension/src/resources/Strings.json:42]: It seems that this variable is not used? The localized string is stored in `package.nls.json`

**Review by tecton — CHANGES_REQUESTED** [review]

File: packages/vscode-extension/src/resources/Strings.json (hunk 16 above)
Comment (wenytang-ms) [reply packages/vscode-extension/src/resources/Strings.json]: removed

File: packages/vscode-extension/src/extension.ts [hunk 17]
```diff
@@ -126,6 +126,27 @@ export async function activate(context: vscode.ExtensionContext) {
   );
   context.subscriptions.push(validateManifestCmd);
 
+  const addApiConnector = vscode.commands.registerCommand(
+    "fx-extension.connectExistingApi",
+    (...args) => Correlator.run(handlers.connectExistingApiHandler, args)
+  );
+
+  const buildPackageCmd = vscode.commands.registerCommand("fx-extension.build", (...args) =>
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply packages/vscode-extension/src/extension.ts]: removed

**Review by tecton — APPROVED** [review]

**Review by jayzhang — APPROVED** [review]

**Pull request merged**